sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1059,3 +1063,282 @@ case class Flatten(child: Expression) extends UnaryExpression {
+      val stop = input2.asInstanceOf[T]
+      val step = input3.asInstanceOf[T]
+
+      var i: Int = getSequenceLength(start, stop, step)
+      val arr = new Array[T](i)
+      while (i > 0) {
+        i -= 1
+        arr(i) = start + step * num.fromInt(i)
+      }
+      arr
+    }
+
+    override def genCode(start: String, stop: String, step: String)
+                        (arr: String, elemType: String)
+                        (implicit ctx: CodegenContext): String = {
+      val i = ctx.freshName("i")
+      s"""
+         |  ${genSequenceLengthCode(start, stop, step)(i)}
+         |  $arr = new $elemType[$i];
+         |  while ($i > 0) {
+         |    $i--;
+         |    $arr[$i] = ($elemType) ($start + $step * $i);
+         |  }
+         """.stripMargin
+    }
+  }
+
+  abstract class TemporalSequence[T: ClassTag](dt: IntegralType, scale: Long, fromLong: Long => T)
+                                              (implicit num: Integral[T])
+    extends SequenceImpl {
+
+    private val backedSequenceImpl = new IntegralSequenceImpl[T]
+    private val microsPerMonth = 28 * CalendarInterval.MICROS_PER_DAY
+
+    override def eval(input1: Any, input2: Any, input3: Any): Array[T] = {
+      val start = input1.asInstanceOf[T]
+      val stop = input2.asInstanceOf[T]
+      val step = input3.asInstanceOf[CalendarInterval]
+      val stepMonths = step.months
+      val stepMicros = step.microseconds
+
+      if (stepMonths == 0) {
+        backedSequenceImpl.eval(start, stop, fromLong(stepMicros / scale))
+
+      } else {
+        // To estimate the resulted array length we need to make assumptions
+        // about a month length in microseconds
+        val minIntervalLengthInMicros = Math.abs(stepMicros) + stepMonths * microsPerMonth
+        val startMicros: Long = num.toLong(start) * scale
+        val stopMicros: Long = num.toLong(stop) * scale
+        val maxEstimatedArrayLength =
+          getSequenceLength(startMicros, stopMicros, minIntervalLengthInMicros)
+
+        val stepSign = if (stopMicros > startMicros) +1 else -1
+        val exclusiveItem = stopMicros + stepSign
+        val arr = new Array[T](maxEstimatedArrayLength)
+        var t = startMicros
+        var i = 0
+
+        while (t < exclusiveItem ^ stepSign < 0) {
+          arr(i) = fromLong(t / scale)
+          t = timestampAddInterval(t, stepMonths, stepMicros, timeZone)
+          i += 1
+        }
+
+        // truncate array to the correct length
+        if (arr.length == i) {
+          arr
+        } else {
+          arr.slice(0, i)
+        }
+      }
+    }
+
+    override def genCode(start: String, stop: String, step: String)
+                        (arr: String, elemType: String)
+                        (implicit ctx: CodegenContext): String = {
+      val stepMonths = ctx.freshName("stepMonths")
+      val stepMicros = ctx.freshName("stepMicros")
+
+      lazy val stepScaled = ctx.freshName("stepScaled")
```

Comment:
Is there any reason to use `lazy` here?

Reply:
No reason. It's a left over, sorry.
Fixed.